File: nearest-exit-from-entrance-in-maze/nearest-exit-from-entrance-in-maze.java
class Solution {
    public int nearestExit(char[][] maze, int[] entrance) {
        int rows = maze.length, cols = maze[0].length;
        int[][] dirs = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        Deque<int[]> q = new ArrayDeque<>();
        int startRow = entrance[0], startCol = entrance[1];
        maze[startRow][startCol] = '+';
        q.offer(new int[]{startRow, startCol, 0});

        while (!q.isEmpty()) {
            int[] currPos = q.poll();
            int currRow = currPos[0], currCol = currPos[1], currDist = currPos[2];

            for (int[] dir : dirs) {
                int nextRow = currRow + dir[0], nextCol = currCol + dir[1];

                if (nextRow >= 0 && nextCol >= 0 && nextRow < rows && nextCol < cols && maze[nextRow][nextCol] == '.') {
                    if (nextRow == 0 || nextRow == rows - 1 || nextCol == 0 || nextCol == cols - 1) {
                        return currDist + 1;
                    }
                    maze[nextRow][nextCol] = '+';
                    q.offer(new int[]{nextRow, nextCol, currDist + 1});
                }
            }
        }

        return -1;
    }
}

File: 2038-nearest-exit-from-entrance-in-maze/nearest-exit-from-entrance-in-maze.py
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        q = deque([entrance])
        m = len(maze)
        n = len(maze[0])
        visited = [[0] * n for _ in range(m)]
        steps = 0
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                #visited[curr[0]][curr[1]] = 1
                if (curr[0] != entrance[0] or curr[1] != entrance[1]) and (curr[0] == 0 or curr[0] == m - 1 or curr[1] == 0 or curr[1] == n - 1):
                    return steps
                for dir in ((1,0), (0,1), (-1,0), (0,-1)):
                    new_pos = (curr[0] + dir[0], curr[1] + dir[1])
                    if 0 <= new_pos[0] < m and 0 <= new_pos[1] < n and maze[new_pos[0]][new_pos[1]] == '.' and not visited[new_pos[0]][new_pos[1]]:
                        q.append(new_pos)
                        visited[new_pos[0]][new_pos[1]] = 1
            steps += 1
        return -1
